### PR TITLE
RD-4070 Fully migrate GenericDeployModal to TypeScript

### PR DESCRIPTION
**Pull request by leon-good-life**
## Description
This pull-request is fully migrates GenericDeployModal to TypeScript. The following line `// @ts-nocheck File not migrated fully to TS` was removed! There are no eslint errors in the terminal.
I have an intention to make more pull requests with further refactoring & enhancements.
I think it will be easier to review not too large pull-requests.

## Screenshots / Videos
N/A

## Checklist

- [x] My code follows the UI Code Style.
- [x] I verified that all tests and checks have passed.
- [x] I verified if that change requires a documentation update.
- [x] I added proper labels to this PR.

## Tests
system tests (evening 06 Feb 2022 - All system tests passed): 
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1983/pipeline

system tests (midday 06 Feb 2022):
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1982/pipeline

System tests:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1972/pipeline

 

## Documentation
N/A

### CODE REVIEW

**Comment by leon-good-life** [comment]
I started system tests: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1972/pipeline

**Review by qooban — COMMENTED** [review]
I'm opting out from the review (at least in this week).
Please work with @kubama .

**Review by jawamartus — COMMENTED** [review]
It looks good for me.
However I'm not enough familiar with our standards and guidelines yet to give approve.

File: widgets/common/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export type Workflow = {
+    name: string;
+    parameters: Record<string, { description?: string; default?: any; type?: any }>;
+    plugin: string;
+    operation?: string;
+};
```
Comment (NorbertAmartus) [inline widgets/common/src/types.ts:6]: As I can see, in the `widgets/common/src/` directory all of the file names are starting with a capital letter.
So to be consistent we should rename the file name to `Types.ts`

Or, alternatively, we could create a subdirectory called `types` in which we'd store a Workflow type as a file (currently there are directories which have that exact purpose e.g `props`)

As we don't have a strict rules about which option should be preferred, I'm leaving it up to you 😉

**Review by NorbertAmartus — COMMENTED** [review]
Looks good 💪 
Added just one comment regarding the file naming/placement

File: widgets/common/src/GenericDeployModal.tsx [hunk 2]
```diff
@@ -4,19 +4,147 @@ import MissingSecretsError from './MissingSecretsError';
...
+    /**
+     * Deployment Name input label
+     */
+    deploymentNameLabel?: string;
+
+    /**
+     * Deployment Name input help description
+     */
+    deploymentNameHelp?: string;
+} & typeof defaultProps;
```
Comment (kubama) [inline widgets/common/src/GenericDeployModal.tsx:119]: Why do we need `& typeof defaultProps` here?

File: widgets/common/src/types.ts (hunk 1 above)
Comment (kubama) [reply widgets/common/src/types.ts:6]: I think a rule is that we are using capital starting letter only for files containing classes or components.

File: widgets/common/src/GenericDeployModal.tsx [hunk 3]
```diff
@@ -4,19 +4,147 @@ import MissingSecretsError from './MissingSecretsError';
 import AccordionSectionWithDivider from './AccordionSectionWithDivider';
 import DeplomentInputsSection from './deployModal/DeploymentInputsSection';
 import DeployModalActions from './deployModal/DeployModalActions';
+import { Workflow } from './types';
```
Comment (kubama) [inline widgets/common/src/GenericDeployModal.tsx:7]: Cosmetic - you can use `import type` here

File: widgets/common/src/GenericDeployModal.tsx [hunk 4]
```diff
@@ -4,19 +4,147 @@ import MissingSecretsError from './MissingSecretsError';
...
+    executeStep: () => void;
+};
 
-    static DEPLOYMENT_SECTIONS = {
-        deploymentInputs: 0,
-        deploymentMetadata: 1,
-        executionParameters: 2,
-        advanced: 3
-    };
+const defaultProps = {
```
Comment (kubama) [inline widgets/common/src/GenericDeployModal.tsx:26]: I can't see the value of this `const` being used anywhere

File: widgets/common/src/GenericDeployModal.tsx [hunk 5]
```diff
@@ -4,19 +4,147 @@ import MissingSecretsError from './MissingSecretsError';
...
-class GenericDeployModal extends React.Component {
-    static EMPTY_BLUEPRINT = { id: '', plan: { inputs: {}, workflows: { install: {} } } };
+type Errors =
+    | {
+          deploymentName?: string;
+          deploymentId?: string;
+          blueprintName?: string;
+          error?: string;
+      }
+    | Record<string, string>;
```
Comment (kubama) [inline widgets/common/src/GenericDeployModal.tsx:19]: Isn't `{
          deploymentName?: string;
          deploymentId?: string;
          blueprintName?: string;
          error?: string;
      }` a subset of `Record<string, string>`?

File: widgets/common/src/GenericDeployModal.tsx (hunk 3 above)
Comment (leon-good-life) [reply widgets/common/src/GenericDeployModal.tsx]: Changed: 777e639

File: widgets/common/src/GenericDeployModal.tsx (hunk 5 above)
Comment (leon-good-life) [reply widgets/common/src/GenericDeployModal.tsx]: Yes, I changed to `Record<string, string>`: 71399d9

File: widgets/common/src/types.ts (hunk 1 above)
Comment (NorbertAmartus) [reply widgets/common/src/types.ts:13]: Ou okay 😁 
But as I'm looking at it right now the [Consts file](https://github.com/cloudify-cosmo/cloudify-stage/blob/master/widgets/common/src/Consts.ts) i just a simple object 🤔 
So I'm not sure if that's just a sign of inconsistency 😁 

If that's the case I'll just resolve this comment 😁

What do you think @kubama?

**Comment by leon-good-life** [comment]
I started system tests:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1982/pipeline

File: widgets/common/src/GenericDeployModal.tsx (hunk 2 above)
Comment (leon-good-life) [reply widgets/common/src/GenericDeployModal.tsx]: I removed it

File: widgets/common/src/GenericDeployModal.tsx (hunk 4 above)
Comment (leon-good-life) [reply widgets/common/src/GenericDeployModal.tsx]: I removed it.

File: widgets/common/src/types.ts (hunk 1 above)
Comment (leon-good-life) [reply widgets/common/src/types.ts:13]: I changed filename to `Types.ts`: 8b0309a

**Comment by leon-good-life** [comment]
I started system tests:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1983/pipeline

File: widgets/common/src/GenericDeployModal.tsx (hunk 4 above)
Comment (leon-good-life) [reply widgets/common/src/GenericDeployModal.tsx]: I made a commit in which I'm using now the defaultProps.

File: widgets/common/src/types.ts (hunk 1 above)
Comment (kubama) [reply widgets/common/src/types.ts:13]: I'm not sure what to think about `Consts.ts`, it exports a single object that is part of `Stage.Common`, so it somehow fits the pattern of the starting case for what's in `Stage.Common`.
For a file containing just types I would say the common approach would be a lower case starting letter:
image

Comment (leon-good-life) [reply widgets/common/src/types.ts:13]: I changed back to `types.ts`: 49ab080

File: widgets/common/src/ExecuteDeploymentModal.tsx [hunk 6]
```diff
@@ -338,7 +338,8 @@ ExecuteDeploymentModal.propTypes = {
     deployments: PropTypes.arrayOf(PropTypes.string),
     workflow: PropTypes.oneOfType([
         PropTypes.shape({ name: PropTypes.string, parameters: PropTypes.shape({}) }),
-        PropTypes.string
+        PropTypes.string,
+        null
```
[the remark this answers is not in the document]
Comment (leon-good-life) [reply widgets/common/src/ExecuteDeploymentModal.tsx]: I changed from:
`PropTypes.oneOfType([ .......  null] )`

to:
`PropTypes.oneOfType([ .......  PropTypes.oneOf([null]) ])`


9ce1e68

File: widgets/common/src/GenericDeployModal.tsx [hunk 7]
```diff
@@ -1,14 +1,161 @@
...
+    showInstallModal: boolean;
+    siteName: string;
+    skipPluginsValidation: boolean;
+    visibility: any;
+    workflow: Workflow;
+};
+
+class GenericDeployModal extends React.Component<GenericDeployModalProps, GenericDeployModalState> {
+    // eslint-disable-next-line react/static-property-placement
+    static defaultProps = defaultProps;
```
[the remark this answers is not in the document]
Comment (leon-good-life) [reply widgets/common/src/GenericDeployModal.tsx:157]: I can't find fully typescript files in cloudify-stage that uses defaultProps. So to declare defaultProps in the class it is what I found online in StackOverflow and other places.

When I move it to the bottom - outside the class, I'm getting an error:
Screen Shot 2022-02-08 at 12 53 02

File: widgets/common/src/ExecuteDeploymentModal.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (leon-good-life) [reply widgets/common/src/ExecuteDeploymentModal.tsx:342]: I reverted my commit back. `null` is a valid prop type. Adding null here fixes error in `npm run check-types`.

**Review by kubama — APPROVED** [review]

**Pull request merged**